File: uni_modules/ui-component/utssdk/app-ios/swift/observer/CoGuestStoreObserver.swift
import Combine
import DCloudUTSFoundation
import RTCRoomEngine_Plus

public class CoGuestStoreObserver {
    private var cancellables = Set<AnyCancellable>()
    public static let shared = CoGuestStoreObserver()
    
    public func coGuestStoreChanged(_ liveId: String, _ callback: @escaping (_ name: String, _ data: String) -> Void) {
        cancellables.removeAll()

        V2RTCRoomEngine.shared.getCoGuestStore(liveId: liveId)
            .state.subscribe(StateSelector(keyPath: \V2CoGuestState.localStatus))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] value in
                guard let self = self else { return }
                callback("localStatus", String(value.rawValue))
            })
            .store(in: &cancellables)

        let arrayKeys: [(String, KeyPath<V2CoGuestState, [SeatUserInfo]>)] = [
            ("invitees", \V2CoGuestState.invitees),
            ("applicants", \V2CoGuestState.applicants),
            ("connectedGuests", \V2CoGuestState.connectedGuests),
            ("invitableGuests", \V2CoGuestState.invitableGuests),
        ]
        for (key, kp) in arrayKeys {
            V2RTCRoomEngine.shared.getCoGuestStore(liveId: liveId)
                .state.subscribe(StateSelector(keyPath: kp))
                .removeDuplicates()
                .receive(on: DispatchQueue.main)
                .sink(receiveValue: { [weak self] arr in
                    guard let self = self else { return }
                    let list = arr.map { TypeConvert.convertSeatUserInfoToDic(userInfo: $0) }
                    if let json = JsonUtil.toJson(list) {
                        callback(key, json)
                    }
                })
                .store(in: &cancellables)
        }

        let requestKeys: [(String, KeyPath<V2CoGuestState, CoGuestRequestInfo?>)] = [
            ("latestReceivedRequest", \V2CoGuestState.latestReceivedRequest),
            ("latestCancelledRequest", \V2CoGuestState.latestCancelledRequest),
            ("latestAcceptedRequest", \V2CoGuestState.latestAcceptedRequest),
            ("latestRejectedRequest", \V2CoGuestState.latestRejectedRequest),
            ("latestTimeoutRequest", \V2CoGuestState.latestTimeoutRequest),
        ]
        for (key, kp) in requestKeys {
            V2RTCRoomEngine.shared.getCoGuestStore(liveId: liveId)
                .state.subscribe(StateSelector(keyPath: kp))
                .removeDuplicates()
                .receive(on: DispatchQueue.main)
                .sink(receiveValue: { [weak self] req in
                    guard let self = self else { return }
                    if let req = req, let json = JsonUtil.toJson(self.convertCoGuestRequestInfoToDic(info: req)) {
                        callback(key, json)
                    } else {
                        callback(key, "")
                    }
                })
                .store(in: &cancellables)
        }
    }

    private func convertCoGuestRequestInfoToDic(info: CoGuestRequestInfo) -> [String: Any] {
        return [
            "requestId": info.requestId,
            "userId": info.userId,
            "userName": info.userName,
            "nameCard": info.nameCard,
            "avatarUrl": info.avatarUrl,
            "timestamp": info.timestamp,
        ]
    }
}

File: uni_modules/ui-component/utssdk/app-android/kotlin/observer/CoGuestStoreObserver.kt
package uts.sdk.modules.uiComponent.observer

import com.google.gson.Gson
import com.tencent.cloud.uikit.state.CoGuestRequestInfo
import com.tencent.cloud.uikit.state.V2RTCRoomEngine
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import uts.sdk.modules.uiComponent.kotlin.observer.TypeConvert

object CoGuestStoreObserver {
    private val gson = Gson()
    private var bindDataJob: Job? = null

    fun coGuestStoreChanged(liveId: String, callback: (String, String) -> Unit) {
        bindDataJob?.cancel()
        bindDataJob = CoroutineScope(Dispatchers.Main).launch {
            launch {
                V2RTCRoomEngine.getInstance().getCoGuestStore(liveId).coGuestState.localStatus.collect { status ->
                    callback("localStatus", gson.toJson(status.value))
                }
            }
            launch {
                V2RTCRoomEngine.getInstance().getCoGuestStore(liveId).coGuestState.invitees.collect { invitees ->
                    val list = invitees?.map { TypeConvert.convertSeatUserInfoToMap(it) }
                    callback("invitees", gson.toJson(list))
                }
            }
            launch {
                V2RTCRoomEngine.getInstance().getCoGuestStore(liveId).coGuestState.applicants.collect { applicants ->
                    val list = applicants?.map { TypeConvert.convertSeatUserInfoToMap(it) }
                    callback("applicants", gson.toJson(list))
                }
            }
            launch {
                V2RTCRoomEngine.getInstance()
                    .getCoGuestStore(liveId).coGuestState.connectedGuests.collect { connectedGuests ->
                        val list = connectedGuests?.map { TypeConvert.convertSeatUserInfoToMap(it) }
                        callback("connectedGuests", gson.toJson(list))
                    }
            }
            launch {
                V2RTCRoomEngine.getInstance()
                    .getCoGuestStore(liveId).coGuestState.invitableGuests.collect { invitableGuests ->
                        val list = invitableGuests?.map { TypeConvert.convertSeatUserInfoToMap(it) }
                        callback("invitableGuests", gson.toJson(list))
                    }
            }

            launch {
                V2RTCRoomEngine.getInstance()
                    .getCoGuestStore(liveId).coGuestState.latestReceivedRequest.collect { request ->
                        callback("latestReceivedRequest", gson.toJson(convertCoGuestRequestInfoToMap(request)))
                    }
            }

            launch {
                V2RTCRoomEngine.getInstance()
                    .getCoGuestStore(liveId).coGuestState.latestCancelledRequest.collect { request ->
                        callback("latestCancelledRequest", gson.toJson(convertCoGuestRequestInfoToMap(request)))
                    }
            }

            launch {
                V2RTCRoomEngine.getInstance()
                    .getCoGuestStore(liveId).coGuestState.latestAcceptedRequest.collect { request ->
                        callback("latestAcceptedRequest", gson.toJson(convertCoGuestRequestInfoToMap(request)))
                    }
            }

            launch {
                V2RTCRoomEngine.getInstance()
                    .getCoGuestStore(liveId).coGuestState.latestRejectedRequest.collect { request ->
                        callback("latestRejectedRequest", gson.toJson(convertCoGuestRequestInfoToMap(request)))
                    }
            }

            launch {
                V2RTCRoomEngine.getInstance()
                    .getCoGuestStore(liveId).coGuestState.latestTimeoutRequest.collect { request ->
                        callback("latestTimeoutRequest", gson.toJson(convertCoGuestRequestInfoToMap(request)))
                    }
            }

        }
    }

    private fun convertCoGuestRequestInfoToMap(request: CoGuestRequestInfo): Map<String, Any> {
        val map = mutableMapOf<String, Any>()
        map["timestamp"] = request.timestamp
        map["requestId"] = request.requestId
        map["userId"] = request.userId
        map["userName"] = request.userName
        map["nameCard"] = request.nameCard
        map["avatarUrl"] = request.avatarUrl
        return map
    }
}